pkg/daemon/daemon.go

```diff
@@ -121,64 +154,78 @@ func (dn *Daemon) Run(stop <-chan struct{}) error {
 		return setUpdateDegraded(dn.kubeClient.CoreV1().Nodes(), dn.name)
 	}
 
+	if isDesired {
+	// we got the machine state we wanted. set the update complete!
+		if err := dn.completeUpdate(); err != nil {
+			glog.Errorf("Marking degraded due to: %v", err)
+			return setUpdateDegraded(dn.kubeClient.CoreV1().Nodes(), dn.name)
+		}
+	} else if err := dn.triggerUpdate(); err != nil {
+		return err
+	}
+
+	if !cache.WaitForCacheSync(stop, dn.nodeListerSynced) {
+		glog.Error("Marking degraded due to: failure to sync caches")
+		return setUpdateDegraded(dn.kubeClient.CoreV1().Nodes(), dn.name)
+	}
+
+	<-stop
+
+	// Run() should block on the above <-stop until an updated is detected,
+	// which is handled by the callbacks.
 	return nil
 }
 
-// process is the main loop that actually does all the work. the flow goes
-// something like this -
-// 1. we restarted for some reason. the happy path reason we restarted is
-//    because of a machine reboot. validate the current machine state is the
-//    desired machine state. if we aren't try updating again. if we are, update
-//    the current state annotation accordingly.
-// 2. watch the desired config annotation, waiting for an update to be
-//    requested by the controller.
-// 3. if an update is requested by the controller, we assume that that means
-//    something changed and apply the desired config no matter what.
-// 4. the update function doesn't return right now, but at some point in the
-//    future if a reboot isn't required for an update it will. if it returns,
-//    validate the machine state and set the update to done.
-//
-// the only reason this function will return is if an error occurs. otherwise it
-// will keep trying to update the machine until it reboots.
-func (dn *Daemon) process() error {
-	for {
-		// validate machine state
-		isDesired, err := dn.isDesiredMachineState()
-		if err != nil {
-			return err
-		}
+// handleNodeUpdate is the handler for informer callbacks detecting node
+// changes. If an update is requested by the controller, we assume that
+// that means something changed and apply the desired config no matter what.
+// Also note that we only care about node updates, not creation or deletion.
+func (dn *Daemon) handleNodeUpdate(old, cur interface{}) {
+	node := cur.(*corev1.Node)
+
+	// First check if the node that was updated is this daemon's node
+	if (node.Name != dn.name) {
+		// The node that was changed was not ours
+		return
+	}
 
-		if isDesired {
-			// we got the machine state we wanted. set the update complete!
-			if err := dn.completeUpdate(); err != nil {
-				return err
-			}
+	// Then check we're not already in a degraded state.
+	if state, err := getNodeAnnotation(dn.kubeClient.CoreV1().Nodes(), dn.name, MachineConfigDaemonStateAnnotationKey); err != nil {
+		// try to set to degraded... because we failed to check if we're degraded
+		glog.Errorf("Marking degraded due to: %v", err)
+		setUpdateDegraded(dn.kubeClient.CoreV1().Nodes(), dn.name)
```

Comment:
I think I'd rather we stop watching and exit on error conditions in this callback function and instead only have the "sleep on Degraded" logic at the beginning of the daemon.

Reply:
Sure, just to make sure, you mean degrade, stop the informer, and then `reboot()` so we can catch it and sleep?

Reply:
Definitely (I thought that was already the case -- when are callbacks actually active? Is it at `nodeInformer.Informer().AddEventHandler()` time? In that case, yeah we should move that to after checking if the node is even degraded.

But also in this function if the node is degraded, let's just stop watching for updates and exit?